### PR TITLE
Implement exact string searching using FM index and LZ index

### PR DESCRIPTION
**Pull request by prolik123**
Added FM Index and unit tests

### CODE REVIEW

**Comment by prolik123** [comment]
Draft! 

Make refactor + write unit tests

File: string_indexing/__pycache__/suffix_array.cpython-310.pyc
Comment (prolik123) [inline string_indexing/__pycache__/suffix_array.cpython-310.pyc:1]: do usuniecia

File: string_indexing/__pycache__/suffix_array.cpython-311.pyc
Comment (prolik123) [inline string_indexing/__pycache__/suffix_array.cpython-311.pyc:1]: do usuniecia

**Review by krzysztof-turowski — CHANGES_REQUESTED** [review]
Thank you, please introduce appropriate changes and let me know if everything is clear.

FYI: `make check` provides you a linter check for all py files, it might be useful to you. In general the convention is as following: `function_name`, `_private_function_name` (i.e. helper, not used outside the package), `variable_name` + `ClassName` (but try to eliminate methods, use it only as C-like structs data holders)

File: string_indexing/fm_index.py [hunk 1]
```diff
@@ -0,0 +1,88 @@
+
+class _FMIndex:
+  def __init__ (self, SA, BWT, text, n):
+    self.L = BWT
+    self.F = '#$' + ''.join(text[SA[i]] for i in range(1, n + 1))
+    self.n = n
+    self.SA = SA
+    self.sample_size = 8 # const for sampling
```
Comment (krzysztof-turowski) [inline string_indexing/fm_index.py:8]: Please make it a class-level constant:
```
class FMIndex:
  SAMPLE_SIZE = 8
```

**Review by krzysztof-turowski — APPROVED** [review]
Thank you very much for your contribution - well done!

**Pull request merged**